set.go

```diff
@@ -678,60 +691,46 @@ func setsFromMsg(msg netlink.Message) (*Set, error) {
 			set.Concatenation = (flags & NFT_SET_CONCAT) != 0
 		case unix.NFTA_SET_KEY_TYPE:
 			nftMagic := ad.Uint32()
-			if invalidMagic, ok := validateKeyType(nftMagic); !ok {
-				return nil, fmt.Errorf("could not determine key type %+v", invalidMagic)
-			}
-			set.KeyType.nftMagic = nftMagic
-			for _, dt := range nftDatatypes {
-				// If this is a non-concatenated type, we can assign the descriptor.
-				if nftMagic == dt.nftMagic {
-					set.KeyType = dt
-					break
-				}
+			dt, err := parseSetDatatype(nftMagic)
+			if err != nil {
+				return nil, fmt.Errorf("could not determine data type: %w", err)
 			}
+			set.KeyType = dt
 		case unix.NFTA_SET_DATA_TYPE:
 			nftMagic := ad.Uint32()
 			// Special case for the data type verdict, in the message it is stored as 0xffffff00 but it is defined as 1
 			if nftMagic == 0xffffff00 {
 				set.KeyType = TypeVerdict
 				break
 			}
-			for _, dt := range nftDatatypes {
-				if nftMagic == dt.nftMagic {
-					set.DataType = dt
-					break
-				}
-			}
-			if set.DataType.nftMagic == 0 {
-				return nil, fmt.Errorf("could not determine data type %x", nftMagic)
+			dt, err := parseSetDatatype(nftMagic)
+			if err != nil {
+				return nil, fmt.Errorf("could not determine data type: %w", err)
 			}
+			set.DataType = dt
 		}
 	}
 	return &set, nil
 }
 
-func validateKeyType(bits uint32) ([]uint32, bool) {
-	var unpackTypes []uint32
-	var invalidTypes []uint32
-	found := false
-	valid := true
-	for bits != 0 {
-		unpackTypes = append(unpackTypes, bits&SetConcatTypeMask)
-		bits = bits >> SetConcatTypeBits
-	}
-	for _, t := range unpackTypes {
-		for _, dt := range nftDatatypes {
-			if t == dt.nftMagic {
-				found = true
-			}
-		}
-		if !found {
-			invalidTypes = append(invalidTypes, t)
-			valid = false
+func parseSetDatatype(magic uint32) (SetDatatype, error) {
+	types := make([]SetDatatype, 0, 32/SetConcatTypeBits)
```

Comment:
This line made me wonder whether cap of 5 types is valid so I researched a bit and found this line in the original source https://git.netfilter.org/nftables/tree/src/datatype.c?id=1acc2fd48c755a8931fa87b8d0560b750316059f#n1165

Code never restricts concat type length to 5 and it is possible to add 6 types via the CLI:
```
$ nft flush ruleset
$ nft add table ip filter
$ nft add set filter setA { type nf_proto . ipv4_addr . inet_service . ipv4_addr . ipv6_addr . ipv6_addr \; flags timeout \; }
$ nft list ruleset
table ip filter {
	set setA {
		type nf_proto . ipv4_addr . inet_service . ipv4_addr . ipv6_addr . ipv6_addr
		flags timeout
	}
}
```

In case there is an overflow, the first type gets either ignored or converted (type is reduced to two bits, e.g. `nf_proto`), meaning that the cap is not specifically checked:
```
$ nft flush ruleset
$ nft add table ip filter
$ # ether_type is 10 so it gets reduced to two bits and converted into nf_proto (2)
$ nft add set filter setB { type ether_type . ipv4_addr . inet_service . ipv4_addr . ipv6_addr . ipv6_addr \; flags timeout \; }
$ nft list ruleset
table ip filter {
	set setB {
		type nf_proto . ipv4_addr . inet_service . ipv4_addr . ipv6_addr . ipv6_addr
		flags timeout
	}
}
```

Now, I am not sure whether using 6 types is a legitimate use case but I see that we check if there is more than 5 types beforehand https://github.com/google/nftables/blob/2729c5a5ee8e93e29b6fe0f857eda704158e19c9/set.go#L197 and with this check, the cap makes sense.

I will leave it up to @konradh to decide if he wants to leave this cap or not.

cc @stapelberg for awareness of this limitation, maybe good to note down in case someone raises an issue in the future.

Reply:
@konradh what do you think?